src/inmanta/data/__init__.py

```diff
@@ -1187,58 +1194,61 @@ class AgentInstance(BaseDocument):
 
     @classmethod
     async def active_for(
-        cls, tid, endpoint, process: uuid.UUID = None, connection: Optional[asyncpg.connection.Connection] = None
-    ):
+        cls: Type[TAgentInstance],
+        tid: uuid.UUID,
+        endpoint: str,
+        process: Optional[uuid.UUID] = None,
+        connection: Optional[asyncpg.connection.Connection] = None,
+    ) -> List[TAgentInstance]:
         if process is not None:
             objects = await cls.get_list(expired=None, tid=tid, name=endpoint, process=process, connection=connection)
         else:
             objects = await cls.get_list(expired=None, tid=tid, name=endpoint, connection=connection)
         return objects
 
     @classmethod
-    async def active(cls):
+    async def active(cls: Type[TAgentInstance]) -> List[TAgentInstance]:
         objects = await cls.get_list(expired=None)
         return objects
 
     @classmethod
-    async def _expire_endpoints_for_process(
-        cls, tid: uuid.UUID, process: uuid.UUID, endpoints: Set[str], now: datetime.datetime
-    ) -> None:
-        """
-        Expire the agent instances for the given endpoints if they are not expired yet.
-        """
-        if not endpoints:
-            return
-        names_subquery = ",".join([f"${i}" for i in range(4, 4 + len(endpoints))])
-        query = f"""UPDATE {cls.table_name()}
-                    SET expired=$1
-                    WHERE tid=$2 AND process=$3 AND name IN ({names_subquery}) AND expired IS NULL
-                 """
-        values = [cls._get_value(now), cls._get_value(tid), cls._get_value(process)] + [cls._get_value(e) for e in endpoints]
-        await cls._execute_query(query, *values)
-
-    @classmethod
-    async def log_instance_creation(
-        cls, tid: uuid.UUID, process: uuid.UUID, endpoints: Set[str], now: datetime.datetime
-    ) -> None:
+    async def log_instance_creation(cls: Type[TAgentInstance], tid: uuid.UUID, process: uuid.UUID, endpoints: Set[str]) -> None:
         """
         Create new agent instances for a given session.
         """
         if not endpoints:
             return
-        # Fix database corruption when database was down
-        await cls._expire_endpoints_for_process(tid, process, endpoints, now)
-        for nh in endpoints:
-            await cls(tid=tid, process=process, name=nh).insert()
+        async with cls.get_connection() as connection:
+            value_tuples: List[Tuple] = [(cls._new_id(), tid, process, name, None) for name in endpoints]
+            value_tuple_len: int = len(value_tuples[0])
+            values_subquery: str = ", ".join(
```

Comment:
This doesn't look right.

I think it should be done using executemanyhttps://magicstack.github.io/asyncpg/current/api/index.html#asyncpg.connection.Connection.executemany

Reply:
I'm pretty sure what I did is correct syntax as well, but your suggestion makes it easier to read: 0b2d7e3e

Reply:
I agree that is is correct syntax (up until the bound of the nr of $'s you can have), but I would assume that it bypasses the query optimizer cache by having a different query every time, while executemany is supported by the postgresql server itself. 

But my assumptions on performance of postgresql have been wrong before.